### PR TITLE
Make Logger optional - fixes #112

### LINKED ISSUE #112: Make Logger optional
Having to create extra object is impediment to getting started with Eliot. And it's not clear Logger is useful in any way.

Current and potential future use cases don't seem super useful:
1. Current: Having tests only validate some messages. This doesn't really make sense - you really want to validate all logged messages in every test.
2. Future: #53, per-Logger destinations. On further thought this doesn't make sense either - if Eliot is logging trees of messages, partial logging on the level of Loggers makes no sense. If you want partial logging it's going to be based on tasks, probably.

### PR DESCRIPTION
**Pull request by itamarst**
Fixes #112.

### CODE REVIEW

File: docs/source/actions.rst [hunk 1]
```diff
@@ -70,11 +68,9 @@ Keep in mind that code within the context block that is run after the action is
 
 .. code-block:: python
 
-     from eliot import start_action, Logger
-
-     logger = Logger()
+     from eliot import start_action
 
-     action = start_action(logger, u"yourapp:subsystem:frob"):
+     action = start_action(action_type=u"yourapp:subsystem:frob"):
```
Comment (itamarst) [inline docs/source/actions.rst]: Need to delete that colon.

File: docs/source/introduction.rst [hunk 2]
```diff
@@ -67,19 +67,16 @@ For simplicity's sake this example focuses on problems 1 and 3; problem 2 is cov
...
   class Person(object):
       def __init__(self):
           self.seen = set()
 
       def look(self, thing):
-          msg = Message.new(type="person:look",
-                            person=unicode(self),
-                            at=unicode(thing))
-          msg.write(logger)
+          Message.new(type="person:look",
```
Comment (itamarst) [inline docs/source/introduction.rst]: Here and later in doc, `message_type` instead of `type`.

**Comment by mithrandi** [comment]
There's a lot of diff here that I didn't read, but I think I like the overall idea. I have a knee-jerk negative reaction to making this implicit instead of explicit, but given the context in which it happens, I think it's fine (it's no less implicit than actions/tasks/etc.). It definitely makes the code easier to write and simplifies the examples; it also seems more likely to lead to code written in a way that can be usefully tested, instead of just punting.

File: docs/source/actions.rst (hunk 1 above)
Comment (wallrj) [inline docs/source/actions.rst]: Which colon?

File: docs/source/types-testing.rst [hunk 3]
```diff
@@ -57,45 +50,43 @@ Here's how we'd test it:
...
             Registration adds entries to the in-memory database.
             """
             registry = UserRegistration()
-            registry.logger = logger
             registry.register(u"john", u"password", 12)
             self.assertEqual(registry.db[u"john"], (u"passsword", 12))
 
 
-Besides calling an the given validation function the ``@validate_logging`` decorator will also validate the logged messages after the test is done.
+Besides calling an the given validation function the ``@capture_logging`` decorator will also validate the logged messages after the test is done.
```
Comment (wallrj) [inline docs/source/types-testing.rst]: Typo "calling an the"

File: docs/source/types-testing.rst [hunk 4]
```diff
@@ -57,45 +50,43 @@ Here's how we'd test it:
...
 You can also call ``MemoryLogger.validate`` yourself to validate written messages.
-If you don't want any additional logging assertions you can decorate your test function using ``@validate_logging(None)``.
+If you don't want any additional logging assertions you can decorate your test function using ``@capture_logging(None)``.
 
 
 Testing Tracebacks
 ------------------
 
-Tests decorated with ``@validate_logging`` will fail if there are any tracebacks logged to the given ``MemoryLogger`` (using ``write_traceback`` or ``writeFailure``) on the theory that these are unexpected errors indicating a bug.
+Tests decorated with ``@capture_logging`` will fail if there are any tracebacks logged to the given ``MemoryLogger`` (using ``write_traceback`` or ``writeFailure``) on the theory that these are unexpected errors indicating a bug.
```
Comment (wallrj) [inline docs/source/types-testing.rst]: :question: "tracebacks logged to the given" > "tracebacks logged to the global"

File: docs/source/types-testing.rst [hunk 5]
```diff
@@ -234,3 +219,93 @@ Or we can simplify further by using ``assertHasMessage`` and ``assertHasAction``
             self.assertEqual(action.children, messages)
             # Each message had the respective server set.
             self.assertEqual(servers, [msg.message["server"] for msg in messages])
+
+
+Restriction Testing to Specific Messages
```
Comment (wallrj) [inline docs/source/types-testing.rst]: typo "Restriction"

File: docs/source/types-testing.rst [hunk 6]
```diff
@@ -234,3 +219,93 @@ Or we can simplify further by using ``assertHasMessage`` and ``assertHasAction``
...
+If you want to only look at certain messages when testing you can log to a specific ``eliot.Logger`` object rather than the global one.
+
+You can log messages to a specific ``Logger``:
+
+.. code-block:: python
+
+    from eliot import Message, Logger
+
+    class YourClass(object):
+        logger = Logger()
```
Comment (wallrj) [inline docs/source/types-testing.rst]: You'd probably want the default logger to be the global logger, right? 
Where do you get that from?

Comment (wallrj) [inline docs/source/types-testing.rst]: Oh, wait...I take that back. Logger() is the global logger....let me read on...

File: eliot/_action.py [hunk 7]
```diff
@@ -189,15 +189,15 @@ def serializeTaskId(self):
...
         """
         Start a new action which is part of a serialized task.
 
         @param logger: The L{eliot.ILogger} to which to write
-            messages.
+            messages, or C{None} if the default one should be used.
 
         @param task_id: A serialized task identifier, the output of
-            L{Action.serialize_task_id}.
+            L{Action.serialize_task_id}. Required.
```
Comment (wallrj) [inline eliot/_action.py]: Well it seems to have a default value of `u""`. Should that be `task_id=_TASK_ID_NOT_SUPPLIED` followed by `if task_id == _TASK_ID_NOT_SUPPLIED: raise Something()` ?

File: docs/source/types-testing.rst (hunk 6 above)
Comment (itamarst) [inline docs/source/types-testing.rst]: Maybe terminology isn't good... the idea is you only need to specify a Logger if you want to restrict scope of messages that show up when testing. Specifying a custom `Logger` is of no benefit or use as far as the non-testing code goes.

File: eliot/_action.py (hunk 7 above)
Comment (itamarst) [inline eliot/_action.py]: Yeah.

File: docs/source/types-testing.rst (hunk 6 above)
Comment (itamarst) [inline docs/source/types-testing.rst]: (My guess is almost no one will ever create a `Logger` ever again once this is merged.)

File: docs/source/actions.rst (hunk 1 above)
Comment (itamarst) [inline docs/source/actions.rst]: At the end of the line.

File: eliot/testing.py [hunk 8]
```diff
@@ -287,6 +288,28 @@ def checkForUnflushed():
...
+
+    See L{validate_logging} for details on the rest of its behavior.
+    """
+    def decorator(function):
+        @validate_logging(assertion, *assertionArgs, **assertionKwargs)
+        @wraps(function)
+        def wrapper(self, *args, **kwargs):
+            logger = kwargs["logger"]
+            current_logger = _output._DEFAULT_LOGGER
+            _output._DEFAULT_LOGGER = logger
```
Comment (wallrj) [inline eliot/testing.py]: Isn't there some (thread related) reason to swap these variables in one statement? eg
`current_logger, _output._DEFAULT_LOGGER = _output._DEFAULT_LOGGER, logger`

File: eliot/testing.py [hunk 9]
```diff
@@ -287,6 +288,28 @@ def checkForUnflushed():
...
+
+            def cleanup():
+                _output._DEFAULT_LOGGER = current_logger
+            self.addCleanup(cleanup)
```
Comment (wallrj) [inline eliot/testing.py]: And if not, perhaps the cleanup function should be registered *before* patching the global logger.

File: eliot/tests/test_action.py [hunk 10]
```diff
@@ -657,6 +657,37 @@ def test_startActionWithParentLogStart(self):
                                   "key": "value"})
 
 
+    def test_startTaskNoLogger(self):
+        """
+        L{startTask} when no logger is given logs to the default ``Logger``.
```
Comment (wallrj) [inline eliot/tests/test_action.py]: :question: Maybe add a comma, or "L{startTask} logs to the default ``Logger`` unless it is supplied with a ``logger`` argument.".

File: eliot/tests/test_action.py [hunk 11]
```diff
@@ -657,6 +657,37 @@ def test_startActionWithParentLogStart(self):
...
+                             {"task_uuid": action._identification["task_uuid"],
+                              "task_level": [1],
+                              "action_type": "sys:do",
+                              "action_status": "started",
+                              "key": "value"})
+
+
+    def test_startActionNoLogger(self):
+        """
+        L{startAction} when no logger is given logs to the default ``Logger``.
```
Comment (wallrj) [inline eliot/tests/test_action.py]: Similar thoughts about this docstring.

File: eliot/tests/test_message.py [hunk 12]
```diff
@@ -72,6 +73,16 @@ def test_writeCallsLoggerWrite(self):
         self.assertEqual(logger.messages[0]["key"], 4)
 
 
+    def test_writeDefaultLogger(self):
+        """
+        L{Message.write} writes the default logger if none is given.
```
Comment (wallrj) [inline eliot/tests/test_message.py]: typo "writes *to* the default logger"

File: eliot/tests/test_testing.py [hunk 13]
```diff
@@ -409,13 +412,13 @@ def test_newMemoryLogger(self):
...
-            @validateLogging(None)
+            @self.validate(None)
             def test_foo(this, logger):
                 result.append(logger)
 
         theTest = MyTest("test_foo")
-        theTest.test_foo()
-        theTest.test_foo()
+        theTest.run()
+        theTest.run()
```
Comment (wallrj) [inline eliot/tests/test_testing.py]: What's the reason for this change to `run()`?

File: eliot/tests/test_testing.py [hunk 14]
```diff
@@ -437,13 +440,13 @@ def test_raises(self):
...
         class MyTest(TestCase):
-            @validateLogging(None)
+            @self.validate(None)
             def test_foo(self, logger):
                 raise exc
 
         raised = None
         try:
-            MyTest("test_foo").test_foo()
+            MyTest("test_foo").debug()
```
Comment (wallrj) [inline eliot/tests/test_testing.py]: Oh, I think I see.

```
In [1]: from unittest import TestCase


In [3]: TestCase.run?
Type:       instancemethod
String Form:<unbound method TestCase.run>
File:       /usr/lib64/python2.7/unittest/case.py
Definition: TestCase.run(self, result=None)
Docstring:  <no docstring>


In [4]: TestCase.debug?
Type:       instancemethod
String Form:<unbound method TestCase.debug>
File:       /usr/lib64/python2.7/unittest/case.py
Definition: TestCase.debug(self)
Docstring:  Run the test without collecting errors in a TestResult

```

But how come this worked before?

File: tox.ini [hunk 15]
```diff
@@ -8,7 +8,7 @@ envlist = pypy, py27, py33, py27-twisted-14, py33-twisted-14,
 commands =
     {envpython} setup.py --version
     pip install eliot[dev]
-    coverage run -m unittest discover eliot.tests
+    coverage run -m unittest discover {posargs:eliot.tests}
```
Comment (wallrj) [inline tox.ini]: Probably something for a follow up issue, but shouldn't we restrict coverage to eliot modules by adding `--source=eliot` and add a `--branches` argument too.

File: docs/source/actions.rst (hunk 1 above)
Comment (wallrj) [inline docs/source/actions.rst]: Ah, I see it now. Yeah do that.

**Comment by wallrj** [comment]
Thanks @itamarst 

This all looks good to me:

 * I had a pretty thorough look through the changes,
 * ran the tests locally - they pass - 
 * built the docs locally  - they build (with a warning `WARNING: html_static_path entry u'/home/richard/projects/HybridLogic/eliot/docs/source/_static' does not exist`)

I asked one or two questions above and noted one or two typos.

Fix / answer those and then merge when you're happy with the build results.

File: eliot/testing.py (hunk 8 above)
Comment (itamarst) [inline eliot/testing.py]: I'm not sure that will actually make it thread-safe. I can just document "testing APIs aren't thread-safe" but that's usually the case.

File: tox.ini (hunk 15 above)
Comment (itamarst) [inline tox.ini]: There's a `.coveragerc` I think.

File: eliot/testing.py (hunk 9 above)
Comment (itamarst) [inline eliot/testing.py]: This is going to blow up no matter what if you use threads.

File: eliot/tests/test_testing.py (hunk 14 above)
Comment (itamarst) [inline eliot/tests/test_testing.py]: I think beforehand no cleanup was necessary; now if you don't cleanup then you have the global logger overriden.
